core/shared/src/main/scala/zio/internal/FiberContext.scala

```diff
@@ -91,19 +91,60 @@ private[zio] final class FiberContext[E, A](
       if (unsafeEvalOn(effect)) ZIO.unit else orElse.unit
     }
 
-  final def getRef[A](ref: FiberRef[A])(implicit trace: ZTraceElement): UIO[A] =
-    ZIO.succeed(unsafeGetRef(ref))
-
   final def id: FiberId.Runtime = fiberId
 
   final def inheritRefs(implicit trace: ZTraceElement): UIO[Unit] = UIO.suspendSucceed {
-    val locals = fiberRefLocals.get
+    val childFiberRefs = fiberRefLocals.get
 
-    if (locals.isEmpty) UIO.unit
+    if (childFiberRefs.isEmpty) UIO.unit
     else
-      UIO.foreachDiscard(locals) { case (fiberRef, value) =>
-        val ref = fiberRef.asInstanceOf[FiberRef[Any]]
-        ref.update(old => ref.join(old, value))
+      ZIO.getFiberRefs.flatMap { parentFiberRefs =>
```

Comment:
Let's move all this logic into `FiberRefs`, to pull it out of the runtime system, as well as make it easier to test.